6.0 NXP-22212 update segment.io connector

yjulienne-nuxeo opened the pull request.

yjulienne-nuxeo committed:
NXP-22212: Use SegmentIO analytics java API version 1.0.7

and js SNIPPET_VERSION 4.0.0

yjulienne-nuxeo committed:
NXP-22212: Refactor Providers to Integrations

yjulienne-nuxeo committed:
NXP-22212: remove openUrls and cache config for segmentIO script

yjulienne-nuxeo committed:
NXP-22212: Set application/javascript content type for segmentIO script

yjulienne-nuxeo committed:
NXP-22212: Do not send first name and last name by default on identify calls

yjulienne-nuxeo committed:
NXP-22212: Allow to send additional traits in client side identifyIfNeeded method

yjulienne-nuxeo committed:
NXP-22212: Cleanup, format and update unit tests

yjulienne-nuxeo committed:
NXP-22212: Force identify call when synchronizing traits

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/nuxeo/nuxeo-segment.io-connector?pullRequest=1) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/nuxeo/nuxeo-segment.io-connector?pullRequest=1) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/nuxeo/nuxeo-segment.io-connector?pullRequest=1) it.</sub>

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/nuxeo/nuxeo-segment.io-connector?pullRequest=1) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/nuxeo/nuxeo-segment.io-connector?pullRequest=1) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/nuxeo/nuxeo-segment.io-connector?pullRequest=1) it.</sub>

nuxeojenkins commented:
From the blame information on this pull request, potential reviewers: @tiry

mincong-h commented on src/main/java/org/nuxeo/segment/io/SegmentIO.java:
```diff
@@ -1,12 +1,12 @@
 package org.nuxeo.segment.io;
```
The license title is missing. Could you help to add it please?

yjulienne-nuxeo replied:
it's 6.0 maintenance branch so I don't bother

mincong-h commented on src/main/java/org/nuxeo/segment/io/SegmentIODataWrapper.java:
```diff
@@ -20,31 +20,28 @@
-            metadata.put("firstName", principal.getFirstName());
-        }
-        if (!metadata.containsKey("lastName")) {
-            metadata.put("lastName", principal.getLastName());
```
Just want to confirm that first-name and last-name are not needed anymore?

yjulienne-nuxeo replied:
I confirm

mincong-h commented on src/main/java/org/nuxeo/segment/io/SegmentIODataWrapper.java:
```diff
@@ -111,8 +103,8 @@ protected boolean isAllowed(Object value) {
-                if (value!=null) {
-                    if ( isAllowed(value)) {
+                if (value != null) {
+                    if (isAllowed(value)) {
```
The code in lines L106-L114 are duplicate with L86-94, please consider if you can refactor them.

yjulienne-nuxeo replied:
yeah, I could, but it's not related to my changes, I would'nt know how to name the method and the gain would be very poor (and I have to port it to 4 branches :D )

mincong-h commented on src/main/java/org/nuxeo/segment/io/SegmentIOIntegrations.java:
```diff
@@ -0,0 +1,20 @@
+package org.nuxeo.segment.io;
```
Please add the missing license, and move the author info from Javadoc to license.

yjulienne-nuxeo replied:
same here (6.0)

mincong-h commented on src/main/resources/META-INF/MANIFEST.MF:
```diff
@@ -1,15 +1,14 @@
+Bundle-RequiredExecutionEnvironment: JavaSE-1.6
+Bundle-Vendor: Nuxeo
+Nuxeo-Component: OSGI-INF/segmentIO-service.xml,OSGI-INF/listeners-con
+ trib.xml
```
Should be the following?

```
Nuxeo-Component: OSGI-INF/segmentIO-service.xml,
  OSGI-INF/listeners-contrib.xml
```

yjulienne-nuxeo replied:
I assume both work

Ronan42 replied:
Wondering why all the lines got modified?

yjulienne-nuxeo replied:
It was an automatic change made somehow by the Nuxeo Eclipse Plugin when I deleted a contribution file

mincong-h commented on src/test/java/org/nuxeo/segment/io/test/TestSegmentIOService.java:
```diff
@@ -46,108 +47,123 @@
+    public void shouldRunIdentifyAndGroupOnLogin() throws Exception {
 
-        SegmentIOComponent component = (SegmentIOComponent) Framework.getRuntime().getComponent(SegmentIOComponent.class.getName());
+        SegmentIOComponent component = (SegmentIOComponent) Framework.getService(SegmentIO.class);
```
I'm wondering if we could use `Framework.getService(SegmentIOComponent.class)` directly?

yjulienne-nuxeo replied:
no, the interface declared as a service is needed

Ronan42 replied:
Should the API needed be actually exposed in the service interface?

yjulienne-nuxeo replied:
it's a getTestData() method made only for testing purpose so I don't think so, but it points to me that I forgot to use the updated interface in SegmentIOAsyncListener ;)

mincong-h reviewed (approved):
:+1:

yjulienne-nuxeo committed:
NXP-22224: Fix groovy classloader memory leak

yjulienne-nuxeo committed:
NXP-22212: Formatting 

+ make logger private 
+ use service interface when possible
+ factoring pushForTest method

mincong-h reviewed (changes requested):
For your commit 8b0cac0, please ensure your Eclipse XML formatter is aligned with the [Nuxeo XML Code Style](https://doc.nuxeo.com/corg/xml-or-xml-like-files-code-style/). Otherwise, it's +1 for me.

Ronan42 commented on src/main/java/org/nuxeo/segment/io/SegmentIOComponent.java:
```diff
@@ -60,7 +61,7 @@
     public final static String MAPPER_EP = "mapper";
 
-    public final static String PROVIDER_EP = "providers";
+    public final static String INTEGRATIONS_EP = "integrations";
```
EP means?

yjulienne-nuxeo replied:
Extension Point

Ronan42 commented on src/main/java/org/nuxeo/segment/io/SegmentIOComponent.java:
```diff
@@ -185,97 +181,102 @@ public void identify(NuxeoPrincipal principal) {
+        return new HashMap<>();
+    }
+
+    protected Options buildOptions() {
```
@since
Javadoc to explain the purpose of this method. Options to do what?

yjulienne-nuxeo replied:
ok

Ronan42 commented on src/main/java/org/nuxeo/segment/io/SegmentIOComponent.java:
```diff
@@ -185,97 +181,102 @@ public void identify(NuxeoPrincipal principal) {
 
-        if (Framework.isTestModeSet()) {
-            pushForTest("identify", wrapper.getUserId(), null, metadata);
+        if (debugMode) {
```
The semantic has changed here. You don't have a test only code, which has been replaced by a debug mode. Is it on purpose?

yjulienne-nuxeo replied:
no, the testMode part has just been moved just beside the "real" API code to allow calling it with exactly the same arguments

Ronan42 commented on src/main/java/org/nuxeo/segment/io/SegmentIOComponent.java:
```diff
@@ -185,97 +181,102 @@ public void identify(NuxeoPrincipal principal) {
-        if (Framework.isTestModeSet()) {
-            pushForTest("identify", wrapper.getUserId(), null, metadata);
+        if (debugMode) {
+            log.info("send identify for " + wrapper.getUserId() + " with meta : " + metadata.toString());
```
boiler plate: if(log.isInfoEnabled()) { ...}

yjulienne-nuxeo replied:
ok

Ronan42 commented on src/main/java/org/nuxeo/segment/io/SegmentIOComponent.java:
```diff
@@ -185,97 +181,102 @@ public void identify(NuxeoPrincipal principal) {
+        SegmentIODataWrapper wrapper = new SegmentIODataWrapper(principal, metadata);
 
         if (!mustTrackprincipal(wrapper.getUserId())) {
             log.debug("Skip user " + principal.getName());
```
if (log.isDebugEnabled()) {}

yjulienne-nuxeo replied:
ok

Ronan42 commented on src/main/java/org/nuxeo/segment/io/SegmentIOComponent.java:
```diff
@@ -185,97 +181,102 @@ public void identify(NuxeoPrincipal principal) {
+        if (debugMode) {
+            log.info("send identify for " + wrapper.getUserId() + " with meta : " + metadata.toString());
+        } else {
+            log.debug("send identify with " + metadata.toString());
```
if (log.isDebugEnabled()) {}

yjulienne-nuxeo replied:
ok

Ronan42 commented on src/main/java/org/nuxeo/segment/io/SegmentIOComponent.java:
```diff
@@ -185,97 +181,102 @@ public void identify(NuxeoPrincipal principal) {
+            Traits traits = new Traits();
+            traits.putAll(wrapper.getMetadata());
+            Options options = buildOptions();
+            if (Framework.isTestModeSet()) {
```
Is this actually used? This 'test' mode?

yjulienne-nuxeo replied:
Yes, in tests :trollface:

Ronan42 commented on src/main/java/org/nuxeo/segment/io/SegmentIOComponent.java:
```diff
@@ -185,97 +181,102 @@ public void identify(NuxeoPrincipal principal) {
 
-    protected void group(String groupId, String userId, Traits traits,
-            Context ctx) {
+    protected void group(String groupId, String userId, Traits traits, Options options) {
```
What does this do, it gets the group descriptor based on input parameters?

Ronan42 replied:
What is the flusher concept?

Ronan42 commented on src/main/java/org/nuxeo/segment/io/SegmentIOComponent.java:
```diff
@@ -185,97 +181,102 @@ public void identify(NuxeoPrincipal principal) {
 
-    protected void pushForTest(String action, String principalName,
-            String eventName, Map<String, Serializable> metadata) {
+    protected void pushForTest(String action, String userId, Traits traits, Options options) {
```
I would rather not have test code together with business code.

Ronan42 replied:
Javadoc to explain what it does.

yjulienne-nuxeo replied:
Yeah, I agree but it would mean mockups in the tests and... maybe for the next time ? ;)

Ronan42 commented on src/main/java/org/nuxeo/segment/io/SegmentIOComponent.java:
```diff
@@ -185,97 +181,102 @@ public void identify(NuxeoPrincipal principal) {
+        testData.add(data);
+    }
+
+    protected void pushForTest(String action, String userId, String eventName, Props eventProperties, Options options) {
```
This should be factorised with the other one.

yjulienne-nuxeo replied:
ok

Ronan42 commented on src/main/java/org/nuxeo/segment/io/SegmentIOComponent.java:
```diff
@@ -297,32 +298,26 @@ public void track(NuxeoPrincipal principal, String eventName) {
+        SegmentIODataWrapper wrapper = new SegmentIODataWrapper(principal, metadata);
 
         if (!mustTrackprincipal(wrapper.getUserId())) {
             log.debug("Skip user " + principal.getName());
```
if (log.isDebugEnabled()) {}

yjulienne-nuxeo replied:
ok

Ronan42 commented on src/main/java/org/nuxeo/segment/io/SegmentIOComponent.java:
```diff
@@ -297,32 +298,26 @@ public void track(NuxeoPrincipal principal, String eventName) {
-        if (Framework.isTestModeSet()) {
-            pushForTest("track", wrapper.getUserId(), eventName, metadata);
+        if (debugMode) {
+            log.info("send track for " + eventName + " user : " + wrapper.getUserId() + " with meta : "
```
if (log.isInfoEnabled()) {}

yjulienne-nuxeo replied:
ok

Ronan42 commented on src/main/java/org/nuxeo/segment/io/SegmentIOComponent.java:
```diff
@@ -297,32 +298,26 @@ public void track(NuxeoPrincipal principal, String eventName) {
-                log.info("send track for " + eventName + " user : "
-                        + wrapper.getUserId() + " with meta : "
-                        + metadata.toString());
+            log.debug("send track with " + metadata.toString());
```
if (log.isDebugEnabled()) {}

yjulienne-nuxeo replied:
ok

Ronan42 commented on src/main/java/org/nuxeo/segment/io/SegmentIODataWrapper.java:
```diff
@@ -60,7 +57,7 @@ public String getUserId() {
     }
 
     // code copied from com.github.segmentio.models.Props
```
? And what is the license of that copied code?

yjulienne-nuxeo replied:
I don't know, and I don't know why we don't use the original method directly, I'll look at it

Ronan42 commented on src/main/java/org/nuxeo/segment/io/SegmentIOIntegrations.java:
```diff
@@ -0,0 +1,20 @@
+
+/**
+ *
+ * @author <a href="mailto:tdelprat@nuxeo.com">Tiry</a>
```
Is this file copied/moved from somewhere else?

yjulienne-nuxeo replied:
Yes, it's a refactoring from SegmentIOProviders.java

Ronan42 commented on src/main/java/org/nuxeo/segment/io/listener/SegmentIOAsyncListener.java:
```diff
@@ -23,16 +23,12 @@
     public void handleEvent(EventBundle bundle) throws ClientException {
 
-        SegmentIOComponent component = getComponent();
+        SegmentIOComponent component = (SegmentIOComponent) Framework.getService(SegmentIO.class);
```
Weird to call a service a component.
And I think there is a slight difference between the two concepts.

yjulienne-nuxeo replied:
👍 forgotten this one thks

Ronan42 commented on src/main/java/org/nuxeo/segment/io/web/SegmentIOScriptResource.java:
```diff
@@ -116,9 +86,8 @@ public Object test() {
-    String email) {
-        if (email==null || email.isEmpty()) {
+    public String getMarketo(@PathParam("email") String email) {
+        if (email == null || email.isEmpty()) {
```
if (isBlank(email)) {

yjulienne-nuxeo replied:
I'm thinking about completely remve this API which should not be used anymore

Ronan42 commented on src/main/resources/skin/views/segmentIOScriptResource/script.ftl:
```diff
@@ -1,35 +1,29 @@
-  window.analytics||(window.analytics=[]),window.analytics.methods=["identify","track","trackLink","trackForm","trackClick","trackSubmit","page","pageview","ab","alias","ready","group","on","once","off"],window.analytics.factory=function(t){return function(){var a=Array.prototype.slice.call(arguments);return a.unshift(t),window.analytics.push(a),window.analytics}};for(var i=0;i < window.analytics ...
-  window.analytics.load("${writeKey}");
-  window.analytics.page();
+  !function(){var analytics=window.analytics=window.analytics||[];if(!analytics.initialize)if(analytics.invoked)window.console&&console.error&&console.error("Segment snippet included twice.");else{analytics.invoked=!0;analytics.methods=["trackSubmit","trackClick","trackLink","trackForm","pageview","identify","reset","group","track","ready","alias","debug","page","once","off","on"];analytics.facto ...
```
This is a very very very long line. Could you format please?

yjulienne-nuxeo replied:
This is the SNIPPET copied directly form SegmentIO website, not sure it brings anything to format it...

Ronan42 commented on src/main/resources/skin/views/segmentIOScriptResource/script.ftl:
```diff
@@ -1,35 +1,29 @@
+      if (document.cookie.indexOf("_nxIdentified="+login)<0 || traits) {
+        // merge email and additional traits
+        if(!traits){
+        	traits = {};
```
format

Ronan42 commented on src/main/resources/skin/views/segmentIOScriptResource/script.ftl:
```diff
@@ -1,35 +1,29 @@
+  }
 <#if principal??>
-   identifyIfNeeded('${principal.name}','${principal.email}');
+	identifyIfNeeded('${principal.name}','${principal.email}');
```
format

Ronan42 commented on src/test/java/org/nuxeo/segment/io/test/TestSegmentIOService.java:
```diff
@@ -46,108 +47,124 @@
+        Assert.assertTrue(allMappers.size() > 0);
         Assert.assertTrue(allMappers.containsKey("loginSuccess"));
-        Assert.assertTrue(allMappers.get("loginSuccess").size()>0);
+        Assert.assertTrue(allMappers.get("loginSuccess").size() > 0);
```
If you feel like it, replace by assertj assertions :-P

Ronan42 commented on src/test/java/org/nuxeo/segment/io/test/TestSegmentIOService.java:
```diff
@@ -46,108 +47,124 @@
+        Assert.assertEquals(2, testData.size());
 
         Map<String, Object> data = testData.remove(0);
+        Assert.assertEquals(6, data.size());
```
use assertj for new assertions.

yjulienne-nuxeo replied:
one day maybe ;)

Ronan42 commented on src/test/java/org/nuxeo/segment/io/test/TestSegmentIOService.java:
```diff
@@ -46,108 +47,124 @@
 
-        Assert.assertTrue(providers.containsKey("Marketo"));
-        Assert.assertTrue((Boolean)providers.get("Marketo"));
+        Assert.assertTrue(integrations.containsKey("Marketo"));
```
useless, the next line would catch this case.

yjulienne-nuxeo replied:
👍

Ronan42 commented on src/test/resources/segmentio-contribs.xml:
```diff
@@ -1,54 +1,58 @@
+			<enableAnonymous>false</enableAnonymous>
+			<blackListedUser>RemoteConnectInstance</blackListedUser>
+		</userFilter>
+	</extension>
```
format the whole file.

yjulienne-nuxeo replied:
ok
